File: python/350.两个数组的交集-ii.py
#
# @lc app=leetcode.cn id=350 lang=python3
#
# [350] 两个数组的交集 II
#
from typing import List

# @lc code=start
class Solution:
    # 排序加双指针
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        len1 = len(nums1)
        len2 = len(nums2)
        index1 = 0
        index2 = 0
        intersection = list()
        while index1 < len1 and index2 < len2:
            if nums1[index1] > nums2[index2]:
                index2 += 1
            elif nums1[index1] < nums2[index2]:
                index1 += 1
            else:
                intersection.append(nums1[index1])
                index1 += 1
                index2 += 1
        return intersection

    # hash表
    def intersect2(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2):
            return self.intersect(nums2, nums1)
        
        m = collections.Counter()

        for i in nums1:
            m[i] += 1

        intersection = list()
        for i in nums2:
            if (counter := m.get(i, 0)) > 0:
                intersection.append(i)
                m[i] -= 1
                if m[i] == 0:
                    m.pop(i)
        return intersection

    # 暴力解法，非常垃圾
    def intersect1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        s1 = set(nums1)
        s2 = set(nums2)
        s = s1 & s2
        countList1 = [0] * len(s)
        countList2 = [0] * len(s)
        for i, num1 in enumerate(s):
            for j in nums1:
                if j == num1:
                    countList1[i] += 1

        for i, num2 in enumerate(s):
            for j in nums2:
                if j == num2:
                    countList2[i] += 1

        maxCount = [0] * len(s)
        for i, _ in enumerate(maxCount):
            maxCount[i] = min(countList1[i], countList2[i])
        
        saveList = list()
        for i, num in enumerate(s):
            saveList.extend([num] * maxCount[i])

        return saveList


# Solution().intersect([4,9,5], [9,4,9,8,4])
# @lc code=end

File: cpp/350.两个数组的交集-ii.cpp
/*
 * @lc app=leetcode.cn id=350 lang=cpp
 *
 * [350] 两个数组的交集 II
 */
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2)
    {
        return method3(nums1, nums2);
    }

    // 排序 + 双指针
    vector<int> method3(vector<int>& nums1, vector<int>& nums2)
    {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        int ptr1 = 0;
        int ptr2 = 0;
        vector<int> ans;
        while (ptr1 < nums1.size() && ptr2 < nums2.size()) {
            if (nums1[ptr1] < nums2[ptr2]) {
                ptr1++;
            }
            else if (nums1[ptr1] > nums2[ptr2]) {
                ptr2++;
            }
            else {
                ans.push_back(nums1[ptr1]);
                ptr1++;
                ptr2++;
            }
        }
        return ans;
    }

    // hash表计数，然后减数。当无法一次性读取所有数据时，可以使用这一方法分批次读取
    vector<int> method2(vector<int>& nums1, vector<int>& nums2)
    {
        if (nums1.size() > nums2.size()) {
            return method2(nums2, nums1);
        }

        unordered_map<int, int> map;

        for (auto i : nums1) {
            map[i]++;
        }

        vector<int> ans;
        for (auto i : nums2) {
            if (map[i]) {
                ans.push_back(i);
                map[i]--;
            }
        }

        return ans;
    }

    // 暴力遍历，思路类似于回溯
    vector<int> method1(vector<int>& nums1, vector<int>& nums2)
    {
        if (nums1.size() > nums2.size()) {
            return method2(nums2, nums1);
        }

        vector<int> ans;
        vector<bool> used(nums2.size(), false);

        for (int i = 0; i < nums1.size(); i++) {
            for (int j = 0; j < nums2.size(); j++) {
                if (nums2[j] == nums1[i] && !used[j]) {
                    ans.push_back(nums1[i]);
                    used[j] = true;
                    break;
                }
            }
        }
        return ans;
    }
};
// @lc code=end